### PR TITLE
Reduce time to first bessely

### LINKED ISSUE #41: Improving compile time of besselj and bessely
I've noticed an increase in compile time for `besselj` and `bessely`
```julia
julia> tstart = time(); using Bessels; tend=time(); tend-tstart
[ Info: Precompiling Bessels [0e736298-9ec6-45e8-9647-e4fc86a2fe38]
0.7326688766479492

julia> tstart = time(); besselk(1.2, 8.2); tend=time(); tend-tstart
0.25269508361816406

julia> tstart = time(); besselj(1.2, 8.2); tend=time(); tend-tstart
1.9875857830047607

julia> tstart = time(); bessely(1.2, 8.2); tend=time(); tend-tstart
0.03860783576965332

julia> tstart = time(); besseli(1.2, 8.2); tend=time(); tend-tstart
0.12631797790527344
```

Loading the package and the calls to `besseli` and `besselk` are ok, however the calls to `bessely` and `besselj` have a noticeable compile time. The reason in this instance that `bessely` is faster is because it shares the majority of its code with `besselj` so the ordering of the call will matter.

This is probably from the asymptotics expansion for large argument in src/asymptotics.jl. I'll look into this some more.

```julia
julia> versioninfo()
Julia Version 1.7.2
Commit bf53498635 (2022-02-06 15:21 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.5.0)
  CPU: Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, westmere)
```

### PR DESCRIPTION
**Pull request by heltonmc**
This works to close #41 and improve compile time. This simply removes the broadcating and adds a precompilation statement to besselj

```julia
julia> tstart = time(); using Bessels; tend=time(); tend-tstart
[ Info: Precompiling Bessels [0e736298-9ec6-45e8-9647-e4fc86a2fe38]
1.4003350734710693

julia> tstart = time(); bessely(1.2, 8.2); tend=time(); tend-tstart
0.6646819114685059

julia> tstart = time(); besselj(1.2, 8.2); tend=time(); tend-tstart
0.07749509811401367
```

This does increasing the package loadtime by about 0.7 seconds but decreases the call to bessely by 1.4 seconds.

One issue with this is these load times are being dominated by a branch that is only hit for a fraction of the arguments. These big branches are only reached if (1.6nu < x < 2nu)... it might be more efficient to use a different algorithm here that converges better and doesn't hurt compilation too much.

Either way I think the structure of this file can be drastically improved to remove redundant compilation.

The current implementation needs to handle the case of Float32 now to make that type stable. It honestly is probably best to do that calculation in higher precision anyway for Float32 numbers considering large argument expansions are prone to more cancellation.

### CODE REVIEW

File: src/asymptotics.jl [hunk 1]
```diff
@@ -121,15 +121,15 @@ function _α_αp_poly_10(v, x::T) where T
...
-    s6 = evalpoly(μ, T.((-5892.0457146167755, 6965.548173427582, -1128.6143367290497, 56.11658954620361, -1.0105445384979248, 0.005837917327880859, -5.0067901611328125e-6)))
-    s7 = evalpoly(μ, T.((231884.6359563172, -276225.9131319225, 46889.68168082833, -2607.1347218453884, 59.29277476668358, -0.5644365847110748, 0.0018794238567352295, -9.834766387939453e-7)))
-    s8 = evalpoly(μ, T.((-1.225106763286615e7, 1.4670878180751745e7, -2.5692206274838205e6, 153362.43350451812, -4001.4928083932027, 49.41718461737037, -0.2788656409829855, 0.0005833245813846588, -1.9976869225502014e-7)))
-    s9 = evalpoly(μ, T.((8.369564497799761e8, -1.0061560971482614e9, 1.80175839935865e8, -1.1293844119552301e7, 322272.3393258663, -4655.44006793818, 34.77845393284224, -0.12591543863527477, 0.0001761710154823959, -4.1618477553129196e-8)))
-    s10 = evalpoly(μ, T.((-7.181611223268327e10, 8.659091218989731e10, -1.576853447634526e10, 1.0243675365394899e9, -3.1127612822052997e7, 499001.19046756154, -4427.281508191307, 21.557856538012857, -0.053067848184582544, 5.209072696743533e-5, -8.843926480039954e-9)))
+    s2 = evalpoly(μ, (-0.1953125, 0.203125, -0.0078125))
```
Comment (devmotion) [inline src/asymptotics.jl:124]: You could keep the conversion to `T` by writing explicitly
```suggestion
    s2 = evalpoly(μ, (T(-0.1953125), T(0.203125), T(-0.0078125)))
```
This is actually what I had in mind - it avoids broadcasting and hence should improve compile time but shouldn't change any results or dispatches.

Comment (heltonmc) [reply src/asymptotics.jl:124]: O interesting! Let me check that out. (or let me see if there is an easy way to automate that 😆)

**Comment by heltonmc** [comment]
The other major structure change is to have the branches within the function so the coefficients themselves are completely reused and we just have to account for varying length tuples.
```julia
function _α_αp_asymptotic(v, x::Float64) where T
    xinv = inv(x)^2
    μ = 4 * v^2
    s0 = one(T)
    s1 = (1 - μ) / 8
    s2 = evalpoly(μ, (-0.1953125, 0.203125, -0.0078125))
    s3 = evalpoly(μ, (1.0478515625, -1.1591796875, 0.1123046875, -0.0009765625))
    s4 = evalpoly(μ, (-11.466461181640625, 13.1358642578125, -1.71624755859375, 0.0469970703125, -0.000152587890625))
    s5 = evalpoly(μ, (211.27614974975586, -246.8455924987793, 37.067405700683594, -1.5151596069335938, 0.017223358154296875, -2.6702880859375e-5))
    s6 = evalpoly(μ, (-5892.0457146167755, 6965.548173427582, -1128.6143367290497, 56.11658954620361, -1.0105445384979248, 0.005837917327880859, -5.0067901611328125e-6))
    s7 = evalpoly(μ, (231884.6359563172, -276225.9131319225, 46889.68168082833, -2607.1347218453884, 59.29277476668358, -0.5644365847110748, 0.0018794238567352295, -9.834766387939453e-7))
    s8 = evalpoly(μ, (-1.225106763286615e7, 1.4670878180751745e7, -2.5692206274838205e6, 153362.43350451812, -4001.4928083932027, 49.41718461737037, -0.2788656409829855, 0.0005833245813846588, -1.9976869225502014e-7))
    s9 = evalpoly(μ, (8.369564497799761e8, -1.0061560971482614e9, 1.80175839935865e8, -1.1293844119552301e7, 322272.3393258663, -4655.44006793818, 34.77845393284224, -0.12591543863527477, 0.0001761710154823959, -4.1618477553129196e-8))
    s10 = evalpoly(μ, (-7.181611223268327e10, 8.659091218989731e10, -1.576853447634526e10, 1.0243675365394899e9, -3.1127612822052997e7, 499001.19046756154, -4427.281508191307, 21.557856538012857, -0.053067848184582544, 5.209072696743533e-5, -8.843926480039954e-9))
    
    if x < 5*v
        s11 = evalpoly(μ, (7.562132599676877e12, -9.139342212003133e12, 1.686263984004851e12, -1.1257747938365613e11, 3.584384568770644e9, -6.188986771037158e7, 616601.7257690447, -3609.808395761912, 12.103919447801218, -0.021179858537379914, 1.5144118606258417e-5, -1.909484126372263e-9))
        s12 = evalpoly(μ, (-9.588236853265692e14, 1.161016591203577e15, -2.164822377227822e14, 1.4768015151140877e13, -4.8748767898914606e11, 8.900147704533855e9, -9.641151011038888e7, 640028.5410983711, -2606.385824929625, 6.277422845836725, -0.008087900592073538, 4.342405588886322e-6, -4.176996526439325e-10))
        s13 = evalpoly(μ, (1.4409993748135704e17, -1.747635582560889e17, 3.286944341694018e16, -2.2818860422499875e15, 7.75248528894739e13, -1.4783938404512664e12, 1.7064572063331423e10, -1.2415577730993801e8, 578046.6774168271, -1705.2186748496474, 3.0504712266430722, -0.0029777336087200013, 1.2309086638850886e-6, -9.237588471933122e-11))
        s14 = evalpoly(μ, (-2.5330554576498356e19, 3.0761765051109855e19, -5.827690025164462e18, 4.1047854971841363e17, -1.4277576770556304e16, 2.819724813029272e14, -3.4214400566025547e12, 2.670056925037507e10, -1.3717353508462998e8, 465890.6725849346, -1028.1881576107903, 1.4039031226739573, -0.0010631539871280848, 3.4554354377824836e-7, -2.061961712485072e-11))
        s15 = evalpoly(μ, (5.149237018485509e21, -6.260437088008504e21, 1.1933270317945127e21, -8.508441473311652e19, 3.0180281284560543e18, -6.1344355915029736e16, 7.750474619361342e14, -6.393706192980096e12, 3.543452904857713e10, -1.3358182185805681e8, 341472.1144099866, -578.7806110577233, 0.6169921485050097, -0.00036976968916281605, 9.619824624385043e-8, -4.639413853091412e-12))
        s16 = evalpoly(μ, (-1.1986350446061028e24, 1.458736721654471e24, -2.795293505867967e23, 2.0139038975089457e22, -7.262922096794037e20, 1.5122633519520676e19, -1.975555191650512e17, 1.7050149800082885e15, -1.0037673890747877e13, 4.1025284625807785e10, -1.1700278886842367e8, 230919.39384718135, -307.22947128726474, 0.260626935759175, -0.0001257334987819726, 2.658906071879852e-8, -1.051117201091023e-12))
        s17 = evalpoly(μ, (3.167717078292175e26, -3.8584266638650446e26, 7.4277307068246446e25, -5.399729057323595e24, 1.9752248974202056e23, -4.197842582814349e21, 5.640147455618286e19, -5.0538188484912346e17, 3.1258357574848365e15, -1.3629748027461303e13, 4.233000035928751e10, -9.36296364559977e7, 145724.23973885347, -155.01792293897955, 0.10637041031529318, -4.191933776256961e-5, 7.303030923530292e-9, -2.3959289142515966e-13))
        s18 = evalpoly(μ, (-9.432675683095812e28, 1.1498154949132059e29, -2.2224123014148547e28, 1.6283472309548362e27, -6.03034763221407e25, 1.3043849059659297e24, -1.7950710383376238e22, 1.6602291470434327e20, -1.0700309670904285e18, 4.92026735148727e15, -1.6365119263912514e13, 3.957355516860315e10, -6.930551343828379e7, 86599.59532013819, -74.82883257079231, 0.042123857348712865, -1.3735762526818516e-5, 1.994760566671595e-9, -5.490670428493242e-14))
        s19 = evalpoly(μ, (3.1436617397720927e31, -3.8346121964186915e31, 7.438172213027661e30, -5.487658176284203e29, 2.0543110479331377e28, -4.512190593153004e26, 6.33949148154985e24, -6.024489803885022e22, 4.020727203383251e20, -1.9328978639294188e18, 6.802550072768408e15, -1.767751414955178e13, 3.3964900375451138e10, -4.792703592867111e7, 48822.71795456563, -34.73905348751307, 0.01624286720956346, -4.43222001918643e-6, 5.421693881235296e-10, -1.2642991118241018e-14))
        s20 = even_odd_poly_add(μ, (-1.1655823449858069e34, 1.4226232333293694e34, -2.7683048917097874e33, 2.0549175959945544e32, -7.766193794571668e30, 1.7289192571208397e29, -2.473355129610242e27, 2.406335820680262e25, -1.6548511836241395e23, 8.261988206301513e20, -3.048956971903377e18, 8.409146358453666e15, -1.7416884938569994e13, 2.7048317602344254e10, -3.1216507081066556e7, 26272.21159305084, -15.578788452840763, 0.0061163849392160095, -1.4107081622121022e-6, 1.4671084930995856e-10, -2.9236916960932355e-15))
        if x < 2*v
            s21 = even_odd_poly_add(μ, (4.7820456670665866e36, -5.839767806978455e36, 1.139609569657776e36, -8.505736729879162e34, 3.241923388548771e33, -7.303696541971675e31, 1.061598557628623e30, -1.0542835823932327e28, 7.4416439961740466e25, -3.8383112218012e23, 1.4749530840065974e21, -4.277160387520148e18, 9.428034079487212e15, -1.582775791680076e13, 2.0161618573658318e10, -1.9279964972155876e7, 13562.78784549061, -6.773533616718505, 0.0022547380647248, -4.4351435924730097e-7, 3.9542564728199e-11, -6.787141437359297e-16))
            s22 = even_odd_poly_add(μ, (-2.1603839510459968e39, 2.639521237493071e39, -5.164169654060501e38, 3.873385769490781e37, -1.4875551809791352e36, 3.3870740530372218e34, -4.993086633044377e32, 5.049452384590897e30, -3.646484825050138e28, 1.934929035289944e26, -7.699788293506606e23, 2.3306500988206443e21, -5.414879939834449e18, 9.69892178268169e15, -1.3389590376401615e13, 1.4169012254830626e10, -1.1355350548005087e7, 6745.941285011119, -2.8643019266046954, 0.0008154154591207295, -1.3789599324701242e-7, 1.061958568923219e-11, -1.5810954484757452e-16))
            s23 = even_odd_poly_add(μ, (1.0699708275991454e42, -1.3078533633276038e42, 2.5647480608865752e41, -1.9322501955238208e40, 7.471528599453178e38, -1.7175047656287374e37, 2.56398527252613e35, -2.6350896428114214e33, 1.9417708082703794e31, -1.0563755426904998e29, 4.333903992576568e26, -1.3614388053852227e24, 3.3089693719445074e21, -6.261128573879847e18, 9.243346554045246e15, -1.0625592945166727e13, 9.4457596395905e9, -6.4083469576330995e6, 3244.624388704194, -1.1811472686414501, 0.0002898188382446864, -4.2443950630387665e-8, 2.84271079789386e-12, -3.694951319807448e-17))
            s24 = even_odd_poly_add(μ, (-5.786059033524315e44, 7.075318200534661e44, -1.3904375539321083e44, 1.051770598467428e43, -4.0921342782576984e41, 9.487906954286141e39, -1.4325415965940442e38, 1.493677625028016e36, -1.1206631946323043e34, 6.2329465039514265e31, -2.626782977025989e29, 8.52403241555426e26, -2.1544180262833415e24, 4.273241060880315e21, -6.678193343397941e18, 8.226793217207085e15, -7.9615210223039795e12, 6.004451011853662e9, -3.4795218728555855e6, 1513.831326319914, -0.4760701870709445, 0.00010139500470517733, -1.2944568154864108e-8, 7.586889731852131e-13, -8.660042155798706e-18))
            s25 = even_odd_poly_add(μ, (3.4037481878842518e47, -4.163722422572948e47, 8.198385202264364e46, -6.22435990386475e45, 2.4353618503436353e44, -5.690731110001809e42, 8.680616408346572e40, -9.169497515863974e38, 6.991529767882026e36, -3.9660440566717967e34, 1.7117897755074633e32, -5.716342746132164e29, 1.495186253411991e27, -3.089698915505606e24, 5.071094031947313e21, -6.62564031827467e18, 6.884682491993269e15, -5.663687113496129e12, 3.6558083666029415e9, -1.8241160174217927e6, 687.0067933018984, -0.1879264018835715, 3.496541491784091e-5, -3.914759734732158e-9, 2.0193378048364477e-13, -2.035109906612696e-18))
            s26 = even_odd_poly_add(μ, (-2.170796074547977e50, 2.6563886780963143e50, -5.239744551853566e49, 3.9915243203539826e48, -1.5697710844957169e47, 3.694226454576932e45, -5.687790414754595e43, 6.07923257088654e41, -4.703231744077508e39, 2.715669597352839e37, -1.1973880447446774e35, 4.101798034247621e32, -1.1059082744278117e30, 2.3689970250205793e27, -4.057770223206878e24, 5.577674944614657e21, -6.157880219725851e18, 5.448993990587677e15, -3.8435237604702446e12, 2.1401262084993436e9, -926134.0995426074, 303.9782133922326, -0.07278018905169194, 1.1898982367956518e-5, -1.1748038677566303e-9, 5.361217221358988e-14, -4.794249299231832e-19))
            s27 = even_odd_poly_add(μ, (1.4962552090981924e53, -1.8315342344152888e53, 3.6186332587139524e52, -2.7651411770117913e51, 1.092593756883467e50, -2.5880088398643682e48, 4.0185482602383215e46, -4.341325948517365e44, 3.40330197791503e42, -1.996788871925338e40, 8.974814912978575e37, -3.1454308144632116e35, 8.712781142577153e32, -1.926817207371682e30, 3.4266421642703216e27, -4.923376108039163e24, 5.727723667496557e21, -5.393876507549732e18, 4.099268265682976e15, -2.498492707776039e12, 1.2086394541382809e9, -456609.83322422515, 131.40817769348385, -0.027695807224792045, 4.000233368100494e-6, -3.500503095889607e-10, 1.4200746208673407e-14, -1.1319755289852937e-19))
            s28 = even_odd_poly_add(μ, (-1.1113660587118494e56, 1.360796163396552e56, -2.692648562464633e55, 2.0634410442908336e54, -8.188692555802185e52, 1.9512412867297912e51, -3.0534326199740573e49, 3.3310997303957132e47, -2.6429272553053182e45, 1.573358131379098e43, -7.195503085038196e40, 2.5742410447166102e38, -7.305470671594484e35, 1.6621868613473995e33, -3.056128428896728e30, 4.5656808004766417e27, -5.560280477035352e24, 5.525982611412455e21, -4.4762143769251487e18, 2.944057191090032e15, -1.561376913130651e12, 6.60436019199837e8, -219122.78889988337, 55.602561558610454, -0.010370022987617815, 1.3297453832052619e-6, -1.0361746982713131e-10, 3.753418608703549e-15, -2.6783349569741323e-20))
            s29 = even_odd_poly_add(μ, (8.871640240590993e58, -1.0865691207568645e59, 2.153044379065757e58, -1.654292598675498e57, 6.591312178536168e55, -1.5792626981209235e54, 2.4890507915218916e52, -2.7398535977577976e50, 2.1978557435213864e48, -1.3258581553903244e46, 6.160064953128667e43, -2.24525395106052e41, 6.512651083776791e38, -1.520126754804397e36, 2.879357904563898e33, -4.453253892004961e30, 5.646779978807957e27, -5.882834513685185e24, 5.036061410449725e21, -3.535410901341911e18, 2.0262400502486065e15, -9.410029770358712e11, 3.500781492364049e8, -102565.9203518178, 23.065519747567176, -0.003825019881917803, 4.374396753566091e-7, -3.048458029283708e-11, 9.900967356259312e-16, -6.3495009755852275e-21))
            s30 = even_odd_poly_add(μ, (-7.592029625744128e61, 9.300807898884993e61, -1.8453653700408117e61, 1.4213662513968141e60, -5.684267643368571e58, 1.3688721521734006e57, -2.1717210753568686e55, 2.410362055956752e53, -1.953160895382593e51, 1.1926329508677187e49, -5.621526725185766e46, 2.084010993835956e44, -6.16598654084696e41, 1.472798231204338e39, -2.8653790379578294e36, 4.5711759648399396e33, -6.008229949414824e30, 6.525582948238488e27, -5.863525845645105e24, 4.3559252317865593e21, -2.6682229171870643e18, 1.3409000748848645e15, -5.4845348768282446e11, 1.8042426705573353e8, -46913.0207873715, 9.39407807634952, -0.0013913710768178739, 1.4251320396079487e-7, -8.917800204478063e-12, 2.6068908049173686e-16, -1.5080064817014915e-21))
            #s31 = evalpoly(μ, (6.94870700249129e64, -8.514689145607698e64, 1.6914454451528807e64, -1.3057856237413843e63, 5.240044268446603e61, -1.2678523824852003e60, 2.023756751680159e58, -2.2633262005265776e56, 1.851154407097591e54, -1.1430278116614541e52, 5.459376169247797e49, -2.0555226619655375e47, 6.192573002831615e44, -1.51046028270532e42, 3.0106526519375556e39, -4.9388963621319927e36, 6.703778092493295e33, -7.556162981360928e30, 7.08676888506171e27, -5.532636463282616e24, 3.5907107782946896e21, -1.9310170558305078e18, 8.557584346416431e14, -3.0990603170297046e11, 9.059551633878309e7, -21002.634410263778, 3.7612116823644834, -0.0004996040659730478, 4.6011577714924795e-8, -2.5949697401648424e-12, 6.851985601150131e-17, -3.587596065338226e-22))
            P = (s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25, s26, s27, s28, s29, s30)
            α, αp = split_phase_evalpoly(xinv, P)
            return α*x, αp
        end
        P = (s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20)
        α, αp = split_phase_evalpoly(xinv, P)
        return α*x, αp
    end
    P = (s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10)
    # solving
    #αp = evalpoly(xinv, (s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13, s14, s15, s16, s17, s18, s19, s20, s21, s22, s23, s24, s25, s26, s27, s28, s29))
    #α = x * (evalpoly(xinv, (s0, -s1, -s2/3, -s3/5, -s4/7, -s5/9, -s6/11, -s7/13, -s8/15, -s9/17, -s10/19, -s11/21, -s12/23, -s13/25, -s14/27, -s15/29, -s16/31, -s17/33, -s18/35, -s19/37, -s20/39, -s21/41, -s22/43, -s23/45, -s24/47, -s25/49, -s26/51, -s27/53, -s28/55, -s29/57)))
    α, αp = split_phase_evalpoly(xinv, P)
    return α*x, αp
end
```
Using this instead will avoid all of the `_α_αp_poly_10`, `_α_αp_poly_20`, `_α_αp_poly_30` which has a bunch of overlapping code.

This would bring the first call down to under a half second without getting rid of any branches or changing the computational details of the code.
```julia
julia> tstart = time(); using Bessels; tend=time(); tend-tstart
[ Info: Precompiling Bessels [0e736298-9ec6-45e8-9647-e4fc86a2fe38]
1.4006271362304688

julia> tstart = time(); bessely(1.2, 8.2); tend=time(); tend-tstart
0.47351694107055664
```

**Comment by heltonmc** [comment]
Ok I'm very hesitant to go with the structure I posted previously because I think it makes it harder to reason about as we add support for Float32 and higher precision calculations. Instead for now I've updated the implementations to convert Float32 to Float64 until I can determine the right cutoffs. In general I think we should do the asymptotic expansions in Float64 no matter what because they are prone to cancellation for large arguments, but I think that discussion should happen in a different PR.

For now, removing the type broadcasts and adding a precompile statement we have gone from startup times.
```julia
# master
ulia> tstart = time(); using Bessels; tend = time(); tend - tstart
[ Info: Precompiling Bessels [0e736298-9ec6-45e8-9647-e4fc86a2fe38]
0.5687751770019531

julia> tstart = time(); bessely(1.2, 8.2); tend = time(); tend - tstart
2.3280398845672607

# this pr
julia> tstart = time(); using Bessels; tend = time(); tend - tstart
[ Info: Precompiling Bessels [0e736298-9ec6-45e8-9647-e4fc86a2fe38]
1.3814761638641357

julia> tstart = time(); bessely(1.2, 8.2); tend = time(); tend - tstart
0.6946120262145996
```

So the startup of `using Bessels` has increased by 0.8 seconds while the time to first call to `bessely` has decreased by 1.65 seconds.

@oscardssmith how do you feel about these changes?

**Pull request merged**